mods/3d_armor/3d_armor/init.lua

```diff
@@ -1,254 +1,514 @@
+				-- The following code returns invalid items to the player's main
+				-- inventory but could open up the possibity for a hacked client
+				-- to receive items back they never really had. I am not certain
+				-- so remove the is_singleplayer check at your own risk :]
+				if minetest.is_singleplayer() and player_inv and
+						player_inv:room_for_item("main", stack) then
+					player_inv:add_item("main", stack)
+				end
+			end
+		end
+	end
+	for item, i in pairs(armor_prev) do
+		local stack = ItemStack(item)
+		-- Previous item is not in current inventory
+		armor:run_callbacks("on_unequip", player, i, stack)
+	end
+end
+
+local function init_player_armor(player)
+	local name = player:get_player_name()
+	local pos = player:get_pos()
+	if not name or not pos then
+		return false
+	end
+	local armor_inv = minetest.create_detached_inventory(name.."_armor", {
+		on_put = function(inv, listname, index, stack, player)
+			validate_armor_inventory(player)
+			armor:save_armor_inventory(player)
+			armor:set_player_armor(player)
+		end,
+		on_take = function(inv, listname, index, stack, player)
+			validate_armor_inventory(player)
+			armor:save_armor_inventory(player)
+			armor:set_player_armor(player)
+		end,
+		on_move = function(inv, from_list, from_index, to_list, to_index, count, player)
+			validate_armor_inventory(player)
+			armor:save_armor_inventory(player)
+			armor:set_player_armor(player)
+		end,
+		allow_put = function(inv, listname, index, put_stack, player)
+			-- DEBUT modif MFF (crabman/24/06/2015)
+			local name = stack:get_name()
+			local player_inv = player:get_inventory()
+			local size = player_inv:get_size(listname)
+			if not ( (name:split(":")[1] == "3d_armor" and stack:get_definition().groups["armor_heal"]) or name:split(":")[1] == "shields") then
+				return 0
+			end
+
+			-- if player class != item class
+			if not pclasses.api.util.can_have_item(player:get_player_name(), name) then
+				return 0
+			end
+
+			-- MFF (crabman/27/11/2015) no same item type. *helmet*
+			local ptype = name:split(":")[2]:split("_")[1]
+			if ptype == "enchanted" then
+			   ptype = name:split(":")[2]:split("_")[2]
+			end
+			for i=1, size do
+				local stack = player_inv:get_stack(listname, i)
+				if stack:get_count() > 0 then
+					if stack:get_name():find(ptype) then
+						return 0
+					end
+				end
+			end
+
+			if name:find("shield") then
+				for i=1, size do
+					local stack = player_inv:get_stack(listname, i)
+					if stack:get_count() > 0 then
+						if armors_no_shields[stack:get_name()] ~= nil then
+							return 0
+						end
+					end
+				end
+			else
+				if armors_no_shields[name] ~= nil then
+					for i=1, size do
+						local stack = player_inv:get_stack(listname, i)
+						if stack:get_count() > 0 then
+							if stack:get_name():find("shields:") then
+								return 0
+							end
+						end
+					end
+				end
+			end
+			-- FIN modif MFF (crabman/24/06/2015)
+			return 1
+		end,
```

Comment:
One of these two `end,` lines should be deleted, otherwise the file doesn't interpret correctly.